TkDQM: Add standalone monitoring modules for OT Rechits and new resolution plots for ITRechits

sroychow opened the pull request:
#### PR description:
This PR adds 

- Modules for OT Rechit monitoring and validation. These can be run only with a standalone config - DQM/SiTrackerPhase2/test/dqmstep_phase2tk_cfg.py. The input to the config should be a GEN-SIM-RECO file. It runs the IT and OT rechit producers together with the DQM & Validation step.
- Fixes an axis range in OT cluster validation module.
- For IT rechits, plots of  cluster res vs cluster size(in X & Y) are added. 
- For IT rechits, a scatter plot of res Y vs res X is added 

#### PR validation:
Validated with wf 23234.0, 34634.0.

cc @gabriel-rmrz @mmusich @tsusa @emiglior

sroychow committed:
add new resolution plots and fix range

sroychow committed:
add standalone modules for monitoring and validation phase2OTrechits

sroychow committed:
code format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22208

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @sroychow (Suvankar Roy Chowdhury) for master.

It involves the following packages:

DQM/SiTrackerPhase2
Validation/SiTrackerPhase2V


@andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please review it and eventually sign? Thanks.
@arossi83, @sroychow this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mmusich commented:
@cmsbuild, please test

mmusich commented on Validation/SiTrackerPhase2V/plugins/Phase2OTValidateRecHit.cc:
```diff
@@ -0,0 +1,766 @@
+      LocalPoint lp = rechitIt->localPosition();
+      //GetSimHits
+      const std::vector<SimHitIdpr>& matchedId = associateRecHit.associateHitId(*rechitIt);
+      //std::cout << "Nmatched SimHits = "  << matchedId.size() << ",";
```
please remove or `LogDebug`

mmusich reviewed (commented):
@sroychow thanks a lot for this!
I left few mostly technical comments below about modernized event and ES data access.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfd2d0/14373/summary.html
**COMMIT**: 1611ce3e0e1ab9a8fe08f64baf0a3820d9dd309e
**CMSSW**: CMSSW_12_0_X_2021-04-20-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33482/14373/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2877046
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 148
 - DQMHistoTests: Total successes: 2876869
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 3668.408 KiB( 37 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 375.258 KiB      TrackerPhase2ITRecHitV/EndCap_Side1
 - DQMHistoSizes: changed ( 23234.0,... ): 375.258 KiB      TrackerPhase2ITRecHitV/EndCap_Side2
 - DQMHistoSizes: changed ( 23234.0,... ): 166.586 KiB      TrackerPhase2ITRecHitV/Barrel
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-20-2300+dfd2d0/42335/triggerResults): no differences found

sroychow committed:
address comments on the PR

sroychow committed:
update other modules to latest syntax of gethandle and esdata

bux fix in accessing handles

simplify product access

update procduct access to recommended format

formatting fix

fix bug in reading digisimlinks

update handle access

sroychow commented:
@mmusich thanks for your comments. Addressed the comments and updated the other modules to the recommended syntax of getting Handle and ES data.

mmusich commented:
> Addressed the comments and updated the other modules to the recommended syntax of getting Handle and ES data.

thank you!

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22216

- This PR adds an extra 52KB to repository

cmsbuild commented:
Pull request #33482 was updated. @andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfd2d0/14382/summary.html
**COMMIT**: f573d48c9cd4630da81f7589d1ebcbf16cfa03b9
**CMSSW**: CMSSW_12_0_X_2021-04-20-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33482/14382/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22218

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #33482 was updated. @andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please check and sign again.

mmusich commented on Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:
```diff
@@ -99,20 +99,23 @@ void Phase2TrackerValidateDigi::dqmBeginRun(const edm::Run& iRun, const edm::Eve
-  iEvent.getByToken(itPixelDigiToken_, itPixelDigiHandle_);
-  iEvent.getByToken(otDigiToken_, otDigiHandle_);
+  //Digis
+  const auto& itPixelDigiHandle_ = iEvent.getHandle(itPixelDigiToken_);
+  const auto& otDigiHandle_ = iEvent.getHandle(otDigiToken_);
+  itdigis_ = itPixelDigiHandle_.product();
+  otdigis_ = otDigiHandle_.product();
```
```suggestion
  itdigis_ =  iEvent.get(itPixelDigiToken_);
  otdigis_ =  iEvent.get(otDigiToken_);
```

mmusich commented on Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:
```diff
@@ -99,20 +99,23 @@ void Phase2TrackerValidateDigi::dqmBeginRun(const edm::Run& iRun, const edm::Eve
   // DigiSimLink
-  iEvent.getByToken(itPixelDigiSimLinkToken_, itPixelSimLinkHandle_);
-  iEvent.getByToken(otDigiSimLinkToken_, otSimLinkHandle_);
+  const auto& itPixelSimLinkHandle_ = iEvent.getHandle(itPixelDigiSimLinkToken_);
+  const auto& otSimLinkHandle_ = iEvent.getHandle(otDigiSimLinkToken_);
+  otSimLink_ = otSimLinkHandle_.product();
+  itSimLink_ = itPixelSimLinkHandle_.product();
```
```suggestion
  otSimLink_ = iEvent.get(itPixelDigiSimLinkToken_);
  itSimLink_ = iEvent.get(otDigiSimLinkToken_);
```
would this work?

sroychow replied:
Please have a look at the update.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22220

- This PR adds an extra 20KB to repository

mmusich reviewed (commented):
Further simplification according to: https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMGetDataFromEvent#get
Nitpicking here.

cmsbuild commented:
Pull request #33482 was updated. @andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please check and sign again.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22222

- This PR adds an extra 56KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22222/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22222/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22223

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #33482 was updated. @andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfd2d0/14391/summary.html
**COMMIT**: cdb48fdba34e80f9164e79b893ef3a64060b23f3
**CMSSW**: CMSSW_12_0_X_2021-04-20-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33482/14391/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2877046
 - DQMHistoTests: Total failures: 3858
 - DQMHistoTests: Total nulls: 149
 - DQMHistoTests: Total successes: 2873017
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 3668.404 KiB( 37 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 375.258 KiB      TrackerPhase2ITRecHitV/EndCap_Side1
 - DQMHistoSizes: changed ( 23234.0,... ): 375.258 KiB      TrackerPhase2ITRecHitV/EndCap_Side2
 - DQMHistoSizes: changed ( 23234.0,... ): 166.586 KiB      TrackerPhase2ITRecHitV/Barrel
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-20-2300+dfd2d0/42347/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22226

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #33482 was updated. @andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please check and sign again.

mmusich commented:
@sroychow there's still a couple of clang-tidy warnings: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfd2d0/14391/llvm-analysis/runCMSClangTidyChecks.log. 
Also since you are at it can you squash the last 8 commits?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-33482/22228

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #33482 was updated. @andrius-k, @kmaeshima, @ErnestaP, @ahmad3213, @cmsbuild, @jfernan2, @rvenditti can you please check and sign again.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-dfd2d0/14401/summary.html
**COMMIT**: c0c9c582946361b69ea1f1e203a6170bd9466bb7
**CMSSW**: CMSSW_12_0_X_2021-04-21-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33482/14401/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2877046
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 148
 - DQMHistoTests: Total successes: 2876863
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 3668.408 KiB( 37 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 375.258 KiB      TrackerPhase2ITRecHitV/EndCap_Side1
 - DQMHistoSizes: changed ( 23234.0,... ): 375.258 KiB      TrackerPhase2ITRecHitV/EndCap_Side2
 - DQMHistoSizes: changed ( 23234.0,... ): 166.586 KiB      TrackerPhase2ITRecHitV/Barrel
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-04-21-1100+dfd2d0/42356/triggerResults): no differences found

jfernan2 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1

davidlange6 commented:
hi @sroychow @cms-sw/dqm-l2 - it looks like this PR adds a dependency on Validation to DQM/SiTrackerPhase2. Could that be fixed? Possibly by moving stuff that needs MC information to Validation. Thanks!

sroychow commented:
@davidlange6 Thanks! should be addressed by #33557. Please have a look.
